### ISSUE TITLE
Using mirageJS with Axios in an existing application

### ISSUE THREAD

**Issue by arctouch-michelhenrique** [miragejs/miragejs#1005]
Hey, I had the idea to use the mirage as hybrid. I mean, to test it better I've created a POC to test the Mirage + Axios. However I'm facing a few problems... it seems the request gone through the mirage but it never reaches the axios response going always to the finally bracket.

I did the POC because I started using in a big project and we I faced it I really did not know how to conduct it.

I saw a few solutions which say to exchange the axios per fetch. But in a big application it could take more time than I expect. Is there a solution to work with both?

server.js

`import { createServer } from "miragejs";

export function makeServer() {
  return new createServer({
    routes() {
      this.passthrough();
      this.passthrough("https://api.github.com/**");

      this.namespace = "/learning";
      this.urlPrefix = "http://localhost:3333";

      this.get("/api/users", () => [
        { id: "1", name: "Luke" },
        { id: "2", name: "Leia" },
        { id: "3", name: "Anakin" },
      ]);
    },
  });
}
`

App.js
`import { useEffect, useState } from "react";
import { api } from "./axios";
import { makeServer } from "./server";

interface UserProps {
  avatar: string;
  blog: string;
  name: string;
}

interface UserFactProps {
  name: string;
  id: number;
}

function App() {
  const [user, setUser] = useState<UserProps>({} as UserProps);
  const [factUser, setFactUser] = useState<UserFactProps[]>([]);
  makeServer();

  useEffect(() => {
    try {
      api.get("/users/michenriq").then((response) => {
        setUser({
          avatar: response.data.avatar_url,
          blog: response.data.blog,
          name: response.data.name,
        });
      });
    } catch (err) {
      console.warn(err);
      console.warn("parou aqui");
    } finally {
      console.log("caiu aqui");
    }

    api.get("http://localhost:3333/learning/api/users").then((response) => {
      setFactUser(response.data);
    });
  }, []);

  return (
    <>
      <h1>{user.name}</h1>
      <p>{user.blog}</p>
      <img src={user.avatar} alt={user.name} />

      {factUser?.map((user) => {
        return <p key={user.id}>{user.name}</p>;
      })}
    </>
  );
}

export default App;
`

axios instance 
ps: I put the interceptors to understand how far the request gone

`import axios, { Axios } from 'axios'

export const api: Axios = axios.create({
  baseURL: 'https://api.github.com'
})

api.interceptors.request.use(function (config) {  
  return config;
}, function (error) {
  console.log(error, 'request error')
  // Do something with request error
  return Promise.reject(error);
});

api.interceptors.response.use(function (response) {
  // Any status code that lie within the range of 2xx cause this function to trigger
  // Do something with response data
  return response;
}, function (error) {
  // Any status codes that falls outside the range of 2xx cause this function to trigger
  // Do something with response error
  return Promise.reject(error);
});`

As the images below the request is made but we never got the response.. 

![Screen Shot 2021-11-08 at 9 26 56 PM](https://user-images.githubusercontent.com/87338530/140839510-f1045ccb-f67c-4a2b-a009-7892634355ab.png)
![Screen Shot 2021-11-08 at 9 27 07 PM](https://user-images.githubusercontent.com/87338530/140839515-93cad633-d8a3-4677-894c-4ddd4efdc4ac.png)

I can do it using hybrid fetch to external api and axios to internal, but I figure it does not the right way to do that.

**Comment by dribeirodev** [comment]
I have the same problem

**Comment by matrunchyk** [comment]
Same issue.

**Comment by michenriq** [comment]
Worked out using Fetch API and mirage@0.1.41 or using Axios and mirage@0.1.41 ?

**Comment by Sheraz064-ai** [comment]
I tried downgrading the mirage version to 0.1.41 but with axios it didn't work.

**Comment by michenriq** [comment]
I get step back three versions, and it looks like does not work well with Axios when you have real APIs and fake APIs at the same page or component.

Sadly I won't change my entire application to use fetch API because it's a lot of work. But I look forward to having this fix if possible in the future.

**Comment by Sheraz064-ai** [comment]
> I get step back three versions, and it looks like does not work well with Axios when you have real APIs and fake APIs at the same page or component.
> 
> Sadly I won't change my entire application to use fetch API because it's a lot of work. But I look forward to having this fix if possible in the future.

Did it work with axios after down grading?

**Comment by arctouch-michelhenrique** [comment]
> > I get step back three versions, and it looks like does not work well with Axios when you have real APIs and fake APIs at the same page or component.
> > Sadly I won't change my entire application to use fetch API because it's a lot of work. But I look forward to having this fix if possible in the future.
> 
> Did it work with axios after down grading?

Yep. The only option I didn't test was downgrading the axios version.

**Comment by Sheraz064-ai** [comment]
> > > I get step back three versions, and it looks like does not work well with Axios when you have real APIs and fake APIs at the same page or component.
> > > Sadly I won't change my entire application to use fetch API because it's a lot of work. But I look forward to having this fix if possible in the future.
> > 
> > 
> > Did it work with axios after down grading?
> 
> Yep. The only option I didn't test was downgrading the axios version.

Can you tell me which version are you using for axios and mirageJS to make it work?

**Comment by arctouch-michelhenrique** [comment]
> > > > I get step back three versions, and it looks like does not work well with Axios when you have real APIs and fake APIs at the same page or component.
> > > > Sadly I won't change my entire application to use fetch API because it's a lot of work. But I look forward to having this fix if possible in the future.
> > > 
> > > 
> > > Did it work with axios after down grading?
> > 
> > 
> > Yep. The only option I didn't test was downgrading the axios version.
> 
> Can you tell me which version are you using for axios and mirageJS to make it work?

It's not working yet. It only works if either it's using axios without real requests mixed with or using FETCH API instead.

**Comment by Sheraz064-ai** [comment]
> > > > > I get step back three versions, and it looks like does not work well with Axios when you have real APIs and fake APIs at the same page or component.
> > > > > Sadly I won't change my entire application to use fetch API because it's a lot of work. But I look forward to having this fix if possible in the future.
> > > > 
> > > > 
> > > > Did it work with axios after down grading?
> > > 
> > > 
> > > Yep. The only option I didn't test was downgrading the axios version.
> > 
> > 
> > Can you tell me which version are you using for axios and mirageJS to make it work?
> 
> It's not working yet. It only works if either it's using axios without real requests mixed with or using FETCH API instead.

It's works for me by downgrading both miragejs and axios version

**Comment by arctouch-michelhenrique** [comment]
> > > > > > I get step back three versions, and it looks like does not work well with Axios when you have real APIs and fake APIs at the same page or component.
> > > > > > Sadly I won't change my entire application to use fetch API because it's a lot of work. But I look forward to having this fix if possible in the future.
> > > > > 
> > > > > 
> > > > > Did it work with axios after down grading?
> > > > 
> > > > 
> > > > Yep. The only option I didn't test was downgrading the axios version.
> > > 
> > > 
> > > Can you tell me which version are you using for axios and mirageJS to make it work?
> > 
> > 
> > It's not working yet. It only works if either it's using axios without real requests mixed with or using FETCH API instead.
> 
> It's works for me by downgrading both miragejs and axios version

Did you downgrade 3 versions of each one?

**Comment by xerardoo** [comment]
I fix the issue downgrading axios to 0.21.0 and works with miragejs  [ 0.1.41, 0.1.42.0, 0.1.43.0 ]

**Comment by ds12will** [comment]
What is the breaking change for the latest version with axios can that breaking change be resolved?

**Comment by Techn1x** [comment]
something to do with `onreadystatechange` in the xhr adapter apparently 
https://stackoverflow.com/questions/69198465/miragejs-cannot-use-axios-if-i-call-api-with-passthrough

**Comment by BryanHunt** [comment]
Any plans to fix this?  Kinda makes Mirage unusable.

**Comment by BryanHunt** [comment]
@IanVS Thanks for the update.
